### ISSUE TITLE
Encrypted instance cannot store strings over 253 characters

### ISSUE THREAD

**Issue by tcK1** [mrousavy/react-native-mmkv#707]
Hey! First of all, thanks for the amazing package.

I'm currently exploring the package within the New Architecture and found the following behavior: When setting strings with over 253 characters, the package does not return the respective value.

For example, here are the values that work and don't work:

Works:
```
[Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*
```

Doesn't work (Added an `+` in the end):
```
[Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*+
```

I've made a [small reproduction repo](https://github.com/tcK1/mmkv-new-arch-issue-repro), but here's the following steps:
- Installed the latest version (When creating this PR it's `3.0.0-beta.7`)
- Enable the New Architecture
- Create a new encrypted storage
- Try to set the string, and change the code between the broken and not string to see the log on the console

It seems that this happens **both on Android and iOS**.

Versions:
`react-native`:  `0.74.3`
`react-native-mmkv`:  `3.0.0-beta.7`

**Comment by mrousavy** [comment]
Hey! 

> Hey! First of all, thanks for the amazing package.

Thank you! :)

> When setting strings with over 253 characters, the package does not return the respective value

Hm, I just tested it with this code:

```ts
try {
  console.log(`setting...`);
  storage.set(
    'dummy_key',
    '[Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*+'
  );
  console.log(`set successfully!`);
  const val = storage.getString('dummy_key')
  console.log(
    `dummy_key length: ${val?.length}; value: ${val}`
  );
} catch (e) {
  console.log(`failed! ${e}`);
}
```

..and this works perfectly fine:

```
 (NOBRIDGE) LOG  setting...
 (NOBRIDGE) LOG  set successfully!
 (NOBRIDGE) LOG  dummy_key length: 254; value: [Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*+
```

**Comment by mrousavy** [comment]
I also just tried to set a 10.000 character string, which works perfectly fine.

Code:

```ts
const storage = new MMKV();

function randomChar(): string {
  const index = Math.floor(Math.random() * 10);
  return 'abcdefghij'[index] as string;
}

const KEY = 'dummy_value';
for (let i = 0; i < 1000; i++) {
  const str = [...Array(i)].map(() => randomChar()).join('');
  console.log(`Setting ${KEY} to a ${i} character string.. (${str})`);
  storage.set(KEY, str);
  const actual = storage.getString(KEY);
  console.log(
    `Returned: ${actual?.length} vs ${str.length}, is equal: ${str === actual}`
  );
}
```

Result:

```
[...998 previous lines]
 (NOBRIDGE) LOG  Setting dummy_value to a 999 character string.. (hiidjiifdabdibbheibafhjjbdhjgdcbcbchbgfacahdjfeehaafehgahgbaghaijdafbedeebbidgbjdbhceijahbfhbdcebbgchdjdbgahcjdebbaibghgcedcgbeifeahejggaeahbjchdbjeedjidbadhfdhfjcffaegbgbfijfhbgebcibcchdefggfgibffdagijdcbajdhgceceeijcaebcbigdcaicehjiajdgaigacfdhcbdedfabcddifhgacbhdjdbicbiicifgcagefabjbijgehacibahgfjjcdiiiifcbcjeddfijajhcibdijacgdebiffccggeebabadfebggiaeiiejjbdicedbfjjgajffiiadchgjhgagicefeiefbbeahgdiiibihiciegheaghddecfibdbcbjbeeaiheajjdhejjbjecdfgcicejdcbicbehddccdccajeaafjagbbfgabgebadceabagdjfdibbeabeejbgbeajafehadegdefiegeeggjdhfjcifjjbhdifiddjdbfddgcaefgfgiaggcddhabaeeajaaidicfaaiabeibbjhgafffjeaahbjiabjbhejhhhjheggjgdbejccggabfhbgicbehbcgdejidbibgchccgceffafhhdbgdiheegaiicchgbjhdcgadhceciijdiabgjbeebfcegbahjhifhgbhfebbbdfadaajjdbiehhagebccbhdcdffegfbdcheidhcaecffefghafdcajgbigdaiadhdaecaghifhgbidhecceeaeafgcahafdjfcbjchijfbihfchjidchbhfefjfifcgjjajifbcadjjicgehagieeajbjcbgfjecjfajaiaaiiicbcehdiaicbfjeggijjcfbcbefhbghajeehijbcgacegchhfbhjccfbjhgahcbejaibhbeeahgjdcfeciaifggdaefgb)
 (NOBRIDGE) LOG  Returned: 999 vs 999, is equal: true
```

**Comment by mrousavy** [comment]
Tested on an iPhone 15 Pro Max simulator, in MMKV Example app on latest main.

**Comment by mrousavy** [comment]
Ah yes you are right. With my latest for loop test this works up to 253, but breaks at 254.

```
 (NOBRIDGE) LOG  Setting dummy_value to a 253 character string.. (ahjidahhfcdfheaiagafbfgefieeafbjjjdcdfegfbideahfdgdhijgjcecdjchgeiejaihjhfifhbffibcdeadgdjhddigbidgjbjbchigaahchbedgigaiagbdcbcejcggfjcjifjbffjjadidhbgdcjhjheggifbgihdgjfiahjiefbafjfebdcfcjibhacgdeiagbfggajbgaihfdecfjdhjbcdfgbgbjhdghdejjhjgfebddgdhighdc)
 (NOBRIDGE) LOG  Returned: 253 vs 253, is equal: true, result: ahjidahhfcdfheaiagafbfgefieeafbjjjdcdfegfbideahfdgdhijgjcecdjchgeiejaihjhfifhbffibcdeadgdjhddigbidgjbjbchigaahchbedgigaiagbdcbcejcggfjcjifjbffjjadidhbgdcjhjheggifbgihdgjfiahjiefbafjfebdcfcjibhacgdeiagbfggajbgaihfdecfjdhjbcdfgbgbjhdghdejjhjgfebddgdhighdc
^ worked

 (NOBRIDGE) LOG  Setting dummy_value to a 254 character string.. (ibjbaefffbeehhhhhbfcjjhebcchdbfgbiijbjjbcbcgeediahgagcigegahciiicfgebicejhgbdebiaaacaegjfdcicbfhjgacjcdibhggajhgdjjhiifjefgcefjbcjagejhbbejfgegjhgcbhhibheccdaffghhbbdafheeaajhbbchghjgaidefebgecicbceabaihgfddddeebefigidbdfgdhagjcaecddebebjgbdijfgifigejgge)
 (NOBRIDGE) LOG  Returned: 2 vs 254, is equal: false, result: M
^ failed
```

But I think this is a Tencent/MMKV issue (or size limitation), I don't think I am doing anything wrong in `set`: https://github.com/mrousavy/react-native-mmkv/blob/b99c51ee0894e149c15eae8934ed48d4ab2237d6/package/cpp/MmkvHostObject.cpp#L107-L110

..and `successful` is even `true`, so it doesn't throw an error here.

Would be great if you can create an issue over at https://github.com/Tencent/MMKV 👍

**Comment by tcK1** [comment]
Yes! Ofc, I will open the issue there as well!

Just to clarify, this behavior only happens with the 3.0.0 release. Using the 2.12.2 (and thus with the New Architecture disabled), it works as expected:

```
 BUNDLE  ./index.js 

 LOG  Value: [Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*+[Qtdiwf-DwGR8#?9rkScC[r&Zzf*5t[6whBz[gaS&9Cd5#JQL*-wQ:8FdSM!0;jvjrA,AkUuKfdxdnEt-e_&gbn9.;&TRZ]gy8[}gx*@1.:wJ;$zT9KiQdLTCMn(Qy!;xebRY:CtjdQynyc.=qY9.,(vT[B1py@_aVyLL?Sji_?0JM(:*/DhrMJ[N!&wS69i9xA,Yjmb[A+u&eN}*N/?&(6?4x_@a{yK6!kGj1H{7WPwG9H9pd?@7,*_yJJ[*+
 LOG  Running "Mmkvnewarch" with {"rootTag":11,"initialProps":{"concurrentRoot":false}}
```

Did we update the MMKV underling version? Maybe I can add this information to the issue over there!

**Comment by mrousavy** [comment]
Interesting - so it's either a recent MMKV upgrade, or maybe the fact that I now use the C++ version of MMKV instead of the Objective-C version.

**Comment by mrousavy** [comment]
Actually on Android I always used the C++ (POSIX) version, on v2 and on v3 - so if the behaviour is the same there (v2 works, v3 breaks), then it's def a recent upgrade.

**Comment by tcK1** [comment]
I've just tested with 2.12.2 on Android and the problem also happens! Seems like this is related to the C++ version then. Thanks for the direction, I will open the issue on their repo!

**Comment by tcK1** [comment]
Opened https://github.com/Tencent/MMKV/issues/1361. 👍🏼

**Comment by mendozammatias** [comment]
Hey! I'm using v2 since I can't enable the new architecture in the project I'm working on. 
Is there any way to publish an update for v2 pointing to the fixed MMKV branch? 
More than happy to collaborate in any possible way.

**Comment by mrousavy** [comment]
Fixed in [react-native-mmkv 3.0.0](https://github.com/mrousavy/react-native-mmkv/releases/tag/v3.0.0)!
